src/c-unchop.c

```diff
@@ -300,6 +446,318 @@ r_obj* list_unchop_fallback(r_obj* ptype,
 
   out = KEEP(vec_slice_fallback(out, locations));
 
-  FREE(7);
+  FREE(8);
+  return out;
+}
+
+static
+r_obj* list_unchop_c_fallback_invoke(
+  r_obj* xs,
+  r_obj* default_,
+  r_obj* name_spec,
+  struct r_lazy error_call
+) {
+  if (default_ != r_null) {
+    xs = push_default(xs, default_);
+  }
+  KEEP(xs);
+
+  r_obj* out = vec_c_fallback_invoke(xs, name_spec, error_call);
+
+  FREE(1);
+  return out;
+}
+
+// This is `vec_c_fallback()` with support for `default`.
+// Keep aligned with `vec_c_fallback()`!
+// TODO: Let `list_unchop()` own fallback helpers and
+// make `vec_c_fallback()` and friends thin wrappers
+// that set `default` to `r_null`, or merge `vec_c()`
+// and `list_unchop()` entirely!
+static
+r_obj* list_unchop_c_fallback(
+  r_obj* ptype,
+  r_obj* xs,
+  r_obj* default_,
+  r_obj* name_spec,
+  const struct name_repair_opts* name_repair,
+  struct vctrs_arg* p_error_arg,
+  struct vctrs_arg* p_default_arg,
+  struct r_lazy error_call
+) {
+  r_obj* cls = KEEP(r_attrib_get(ptype, syms_fallback_class));
+  bool implements_c = class_implements_base_c(cls);
+
+  if (implements_c) {
+    FREE(1);
+    return list_unchop_c_fallback_invoke(xs, default_, name_spec, error_call);
+  }
+
+  struct fallback_opts fallback_opts = {
+    .s3 = S3_FALLBACK_false
+  };
+
+  struct ptype_common_opts ptype_opts = {
+    .p_arg = p_error_arg,
+    .call = error_call,
+    .fallback = fallback_opts
+  };
+
+  // Should cause a common type error, unless another fallback
+  // kicks in (for instance, homogeneous class with homogeneous
+  // attributes)
+  ptype = KEEP(vec_ptype_common_opts(xs, r_null, &ptype_opts));
+
+  if (default_ != r_null) {
+    // Also check against `default` using `p_default_arg`
+    const struct ptype2_opts ptype2_opts = {
+      .x = ptype,
+      .y = default_,
+      .p_x_arg = vec_args.empty,
+      .p_y_arg = p_default_arg,
+      .call = error_call,
+      .fallback = fallback_opts
+    };
+    int _;
+    vec_ptype2_opts(&ptype2_opts, &_);
```

Comment:
Very large amount of effort went in to making sure we check the ptype against `default` here as well in a way that results in good errors using `p_default_arg`.

Can't just blindly merge `default` into `xs` ahead of time, because we want to use `p_default_arg` when type checking `default`, and `p_error_arg` when checking `xs`.

We have snapshot tests to ensure we keep emitting good errors in these weird fallback cases.